File: spring-boot-rest-services-java/src/main/java/de/larmic/rest/rest/TweetController.java
package de.larmic.rest.rest;

import de.larmic.rest.rest.dto.TweetDto;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Collection;
import java.util.UUID;

@RestController
public class TweetController {

    private final InMemoryRepository inMemoryRepository;

    public TweetController(final InMemoryRepository inMemoryRepository) {
        this.inMemoryRepository = inMemoryRepository;
    }

    @PostMapping("/")
    public TweetDto tweet(@RequestBody(required = true) final String message) {
        final var dto = new TweetDto(UUID.randomUUID().toString(), message);
        this.inMemoryRepository.tweets.put(dto.id(), dto);
        return dto;
    }

    @GetMapping("/{id}")
    public ResponseEntity<TweetDto> readTweet(@PathVariable final String id) {
        if (this.inMemoryRepository.tweets.containsKey(id)) {
            return ResponseEntity.ok(this.inMemoryRepository.tweets.get(id));
        }

        return ResponseEntity.notFound().build();
    }

    @GetMapping("/")
    public Collection<TweetDto> readAllTweets() {
        return this.inMemoryRepository.tweets.values();
    }

    @PutMapping("/{id}")
    public ResponseEntity<TweetDto> updateTweet(@PathVariable final String id, @RequestBody String message) {
        if (this.inMemoryRepository.tweets.containsKey(id)) {
            final var dto = new TweetDto(id, message);
            this.inMemoryRepository.tweets.put(id, dto);

            return ResponseEntity.ok(dto);
        }

        return ResponseEntity.notFound().build();
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteTweet(@PathVariable final String id) {
        if (this.inMemoryRepository.tweets.containsKey(id)) {
            this.inMemoryRepository.tweets.remove(id);
            return ResponseEntity.ok().build();
        }

        return ResponseEntity.notFound().build();
    }
}

File: spring-boot-rest-services-kotlin/src/main/kotlin/de/larmic/rest/rest/TweetController.kt
package de.larmic.rest.rest

import de.larmic.rest.rest.dto.TweetDto
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.*

@RestController
class TweetController(private val inMemoryRepository: InMemoryRepository) {

    @PostMapping("/")
    fun tweet(@RequestBody(required = true) message: String) = message.wrapInTweet().storeInDatabase()

    @GetMapping("/{id}")
    fun readTweet(@PathVariable id: String): ResponseEntity<TweetDto?> = if (inMemoryRepository.tweets.containsKey(id)) {
        ResponseEntity.ok(inMemoryRepository.tweets[id])
    } else ResponseEntity.notFound().build()

    @GetMapping("/")
    fun readAllTweets() = inMemoryRepository.tweets.values

    @PutMapping("/{id}")
    fun updateTweet(@PathVariable id: String, @RequestBody message: String): ResponseEntity<TweetDto> {
        if (inMemoryRepository.tweets.containsKey(id)) {
            return message.wrapInTweet().storeInDatabase().wrapInResponse()
        }
        return ResponseEntity.notFound().build()
    }

    @DeleteMapping("/{id}")
    fun deleteTweet(@PathVariable id: String): ResponseEntity<*> {
        if (inMemoryRepository.tweets.containsKey(id)) {
            inMemoryRepository.tweets.remove(id)
            return ResponseEntity.ok().build<Any>()
        }
        return ResponseEntity.notFound().build<Any>()
    }

    private fun String.wrapInTweet() = TweetDto(UUID.randomUUID().toString(), this)

    private fun TweetDto.storeInDatabase(): TweetDto {
        inMemoryRepository.tweets[this.id] = this
        return this
    }

    private fun TweetDto.wrapInResponse() = ResponseEntity.ok(this)
}